File: lang/ruby/test/fun/vac/data_structures/CircularlyLinkedListTest.rb
require "fun/vac/data_structures/CircularlyLinkedList"
require "fun/vac/util/TestRunner"

def testCircularlyLinkedList()
    size = 8192

    list = CircularlyLinkedList::CircularlyLinkedList.new()

    for i in 1 .. size
        list.append(i)
    end

    if list.size() != size
        return false
    end

    for i in 0 ... size
        if list.get(i) != i + 1
            return false
        end
    end

    for i in 0 ... size
        list.set(i, size - i)
    end

    for i in 0 ... size
        if list.get(i) != size - i
            return false
        end
    end

    x, y = 0, 0

    i, j = 0, size - 1
    while i < j
        x = list.get(i)
        y = list.get(j)

        list.remove(i)
        list.insert(i, y)
        list.remove(j)
        list.insert(j, x)
        i, j = i + 1, j - 1
    end

    for i in 0 ... size
        if list.get(i) != i + 1
            return false
        end
    end

    i = size
    while i >= 1
        if list.back() != i
            return false
        end
        list.eject()
        i -= 1
    end

    unless list.isEmpty()
        return false
    end

    return true
end

if __FILE__ == $0
    TestRunner.parseTest(testCircularlyLinkedList())
end

File: lang/java/test/fun/vac/data_structures/CircularlyLinkedListTest.java
import fun.vac.data_structures.CircularlyLinkedList;
import fun.vac.util.TestRunner;

public class CircularlyLinkedListTest {

    public static boolean testCircularlyLinkedList() {
        int size = 8192;

        CircularlyLinkedList<Integer> list = new CircularlyLinkedList<>();

        for (int i = 1; i <= size; i++) {
            list.append(i);
        }

        if (list.size() != size) {
            return false;
        }

        for (int i = 0; i < size; i++) {
            if (list.get(i) != i + 1) {
                return false;
            }
        }

        for (int i = 0; i < size; i++) {
            list.set(i, size - i);
        }

        for (int i = 0; i < size; i++) {
            if (list.get(i) != size - i) {
                return false;
            }
        }

        int x;
        int y;

        for (int i = 0, j = size - 1; i < j; i++, j--) {
            x = list.get(i);
            y = list.get(j);

            list.remove(i);
            list.insert(i, y);
            list.remove(j);
            list.insert(j, x);
        }

        for (int i = 0; i < size; i++) {
            if (list.get(i) != i + 1) {
                return false;
            }
        }

        for (int i = size; i >= 1; i--) {
            if (list.back() != i) {
                return false;
            }
            list.eject();
        }

        return list.isEmpty();
    }

    public static void main(String[] args) {
        TestRunner.parseTest(testCircularlyLinkedList());
    }
}
